beacon-chain/core/epoch/epoch_operations.go

```diff
@@ -0,0 +1,338 @@
+			winnerBalance = rootBalance
+			winnerRoot = candidateRoot
+		}
+	}
+	return winnerRoot, nil
+}
+
+// AttestingValidators returns the validators of the winning root.
+//
+// Spec pseudocode definition:
+//    Let `attesting_validators(shard_committee)` be equal to
+//    `attesting_validator_indices(shard_committee, winning_root(shard_committee))` for convenience
+func AttestingValidators(
+	state *pb.BeaconState,
+	shardCommittee *pb.ShardAndCommittee,
+	thisEpochAttestations []*pb.PendingAttestationRecord,
+	prevEpochAttestations []*pb.PendingAttestationRecord) ([]uint32, error) {
+
+	root, err := WinningRoot(
+		state,
+		shardCommittee,
+		thisEpochAttestations,
+		prevEpochAttestations)
+	if err != nil {
+		return nil, fmt.Errorf("could not get winning root: %v", err)
+	}
+
+	indices, err := validators.AttestingValidatorIndices(
+		state,
+		shardCommittee,
+		root,
+		thisEpochAttestations,
+		prevEpochAttestations)
+	if err != nil {
+		return nil, fmt.Errorf("could not get attesting validator indices: %v", err)
+	}
+
+	return indices, nil
+}
+
+// TotalAttestingBalance returns the total balance at stake of the validators
+// attested to the winning root.
+//
+// Spec pseudocode definition:
+//    Let total_balance(shard_committee) =
+//    sum([get_effective_balance(state, i) for i in shard_committee.committee])
+func TotalAttestingBalance(
+	state *pb.BeaconState,
+	shardCommittee *pb.ShardAndCommittee,
+	thisEpochAttestations []*pb.PendingAttestationRecord,
+	prevEpochAttestations []*pb.PendingAttestationRecord) (uint64, error) {
+
+	var totalBalance uint64
+	attestedValidatorIndices, err := AttestingValidators(state, shardCommittee, thisEpochAttestations, prevEpochAttestations)
+	if err != nil {
+		return 0, fmt.Errorf("could not get attesting validator indices: %v", err)
+	}
+
+	for _, index := range attestedValidatorIndices {
+		totalBalance += validators.EffectiveBalance(state, index)
+	}
+
+	return totalBalance, nil
+}
+
+// TotalBalance returns the total balance at stake of the validators
+// from the shard committee regardless of validators attested or not.
+//
+// Spec pseudocode definition:
+//    Let total_balance(shard_committee) =
+//    sum([get_effective_balance(state, i) for i in shard_committee.committee])
+func TotalBalance(
+	state *pb.BeaconState,
+	shardCommittee *pb.ShardAndCommittee) uint64 {
+
+	var totalBalance uint64
+	for _, index := range shardCommittee.Committee {
+		totalBalance += validators.EffectiveBalance(state, index)
+	}
+
+	return totalBalance
+}
+
+// InclusionSlot returns the slot number of when the validator's
+// attestation gets included in the beacon chain.
+//
+// Spec pseudocode definition:
+//    Let inclusion_slot(state, index) =
+//    a.slot_included for the attestation a where index is in
+//    get_attestation_participants(state, a.data, a.participation_bitfield)
+func InclusionSlot(state *pb.BeaconState, validatorIndex uint32) (uint64, error) {
```

Comment:
shouldn't this take an attestation as an argument ?

Reply:
Nope, because we pass in `state`. Using the `validatorIndex` it finds the `attestation` in state

Reply:
What is the bound on latest attestations, are they updated each epoch ? Otherwise I could see multiple inclusion slots for a single validator

Reply:
Each validator only has one inclusion slot per epoch. If a irrational validator creates 2 attestations per epoch. One is valid and one is invalid. The invalid one won't even get to state's `pendingAttestation` list, it'd be rejected during block processing and the entire block would be discarded.

Reply:
Sounds good